### PR TITLE
jsonnet: add tsdb config in jsonnet

### LINKED ISSUE #9789: Jsonnet support configuring tsdb storage
**Is your feature request related to a problem? Please describe.**
Current implementation of jsonnet build does not support configuring tsdb storage as  stateful compactors are created only`using_boltdb_shipper` is true.

**Describe the solution you'd like**
Add new variable `using_tsdb_shipper`, when its true create stateful compactors. Rename `boltdb_shipper_shared_store` to `shared_store`.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by manohar-koukuntla**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #9789 

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
  - [x] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`
- [x] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

**Comment by manohar-koukuntla** [comment]
Loki did not upgrade their jsonnet deployments for tsdb support, created a PR on loki to add it
https://github.com/grafana/loki/pull/9790

File: production/ksonnet/loki/shipper.libsonnet [hunk 1]
```diff
@@ -8,48 +8,60 @@
   local containerPort = k.core.v1.containerPort,
 
   _config+:: {
+    // flag for tuning things when boltdb-shipper is current or upcoming index type.
+    using_boltdb_shipper: true,
+    using_tsdb_shipper: false,
+    using_store: $._config.using_boltdb_shipper || $._config.using_tsdb_shipper,
```
Comment (chaudum) [inline production/ksonnet/loki/shipper.libsonnet:14]:
```suggestion
    using_shipper_store: $._config.using_boltdb_shipper || $._config.using_tsdb_shipper,
```

File: production/ksonnet/loki/shipper.libsonnet [hunk 2]
```diff
@@ -8,48 +8,60 @@
   local containerPort = k.core.v1.containerPort,
 
   _config+:: {
+    // flag for tuning things when boltdb-shipper is current or upcoming index type.
+    using_boltdb_shipper: true,
+    using_tsdb_shipper: false,
```
Comment (chaudum) [inline production/ksonnet/loki/shipper.libsonnet:13]: I think we can also enable this flag by default. Wdyt?

File: CHANGELOG.md [hunk 3]
```diff
@@ -102,6 +102,7 @@
 
 #### Jsonnet
 
+* [9790](https://github.com/grafana/loki/pull/9790) **manohar-koukuntla**: Jsonnet config update for tsdb.
```
Comment (chaudum) [inline CHANGELOG.md:105]: nit:

```suggestion
* [9790](https://github.com/grafana/loki/pull/9790) **manohar-koukuntla**: Add TSDB equivalent of `use_boltdb_shipper` flag to be able to configure `tsdb_shipper` section.
```

File: production/ksonnet/loki/shipper.libsonnet (hunk 2 above)
Comment (manohar-koukuntla) [reply production/ksonnet/loki/shipper.libsonnet:13]: I think its difficult to enable this by default as we also need to add schema config with a specific date, when they want to move to tsdb.

`
      schema_config: {
        configs: [{
          from: '2023-01-05',
          index: {
            period: '24h',
            prefix: 'loki_index_',
          },
          object_store: 'azure',
          schema: 'v12',
          store: 'boltdb-shipper',
        }, {
          from: '2023-06-27',
          index: {
            period: '24h',
            prefix: 'loki_index_',
          },
          object_store: 'azure',
          schema: 'v12',
          store: 'tsdb',
        }],
      },
`

Comment (chaudum) [reply production/ksonnet/loki/shipper.libsonnet:13]: Fair point. I'm also ok to leave it as it.

**Review by chaudum — APPROVED** [review]

**Pull request merged**